Sources/MongoSwift/Operations/UpdateOperation.swift

```diff
@@ -0,0 +1,134 @@
+    public let collation: Document?
+
+    /// When true, creates a new document if no document matches the query.
+    public let upsert: Bool?
+
+    /// An optional WriteConcern to use for the command.
+    public let writeConcern: WriteConcern?
+
+    /// Convenience initializer allowing any/all parameters to be optional
+    public init(arrayFilters: [Document]? = nil,
+                bypassDocumentValidation: Bool? = nil,
+                collation: Document? = nil,
+                upsert: Bool? = nil,
+                writeConcern: WriteConcern? = nil) {
+        self.arrayFilters = arrayFilters
+        self.bypassDocumentValidation = bypassDocumentValidation
+        self.collation = collation
+        self.upsert = upsert
+        self.writeConcern = writeConcern
+    }
+}
+
+/// Options to use when executing a `replaceOne` command on a `MongoCollection`.
+public struct ReplaceOptions: UpdateOperationOptions {
+    /// If true, allows the write to opt-out of document level validation.
+    public let bypassDocumentValidation: Bool?
+
+    /// Specifies a collation.
+    public let collation: Document?
+
+    /// When true, creates a new document if no document matches the query.
+    public let upsert: Bool?
+
+    /// An optional `WriteConcern` to use for the command.
+    public let writeConcern: WriteConcern?
+
+    /// Convenience initializer allowing any/all parameters to be optional
+    public init(bypassDocumentValidation: Bool? = nil,
+                collation: Document? = nil,
+                upsert: Bool? = nil,
+                writeConcern: WriteConcern? = nil) {
+        self.bypassDocumentValidation = bypassDocumentValidation
+        self.collation = collation
+        self.upsert = upsert
+        self.writeConcern = writeConcern
+    }
+}
+
+/// An operation corresponding to an `update` command on a collection.
+internal struct UpdateOperation<CollectionType: Codable, OptionsType: UpdateOperationOptions>: Operation {
+    private let collection: MongoCollection<CollectionType>
+    private let filter: Document
+    private let update: Document
+    private let options: OptionsType?
+    private let type: UpdateType
+
+    internal enum UpdateType {
+        case updateOne, updateMany, replaceOne
+    }
+
+    internal init(collection: MongoCollection<CollectionType>,
+                  filter: Document,
+                  update: Document,
+                  options: OptionsType?,
+                  type: UpdateType) {
+        self.collection = collection
+        self.filter = filter
+        self.update = update
+        self.options = options
+        self.type = type
+    }
+
+    internal func execute() throws -> UpdateResult? {
+        let opts = try self.collection.encoder.encode(self.options)
+        let reply = Document()
+        var error = bson_error_t()
+
+        var result: Bool!
+        switch self.type {
+        case .updateOne:
+            result = mongoc_collection_update_one(
```

Comment:
I did try the approach of defining a lookup map here that would map an `UpdateType` to a libmongoc function which we would retrieve and then call, since all the parameters we pass in are identical and then we could just do something like `result = updateFunc[self.type]!(params...)`.

however, that provided challenging:
- genericized types are not allowed to have static properties, so I could not define the map in here.
- I tried just having it free-floating, but I can't reference the nested `UpdateType` as `UpdateOperation.UpdateType` - it needs to be `UpdateOperation<Type1, Type2>.UpdateType`... which is strange, we'd have to pick arbitrary concrete types, even though they aren't actually relevant to the `UpdateType`. 

if anyone feels strongly I could move `UpdateType` to not be nested, but I do feel like it belongs within the operation type.

Reply:
You could make the enum cases associated with `Int`s and make the map `[Int: <libmongoc function type>]`, then index by the cases raw value.